onnxruntime/core/providers/qnn/builder/opbuilder/cast_op_builder.cc

```diff
@@ -0,0 +1,138 @@
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace qnn {
+
+class CastOpBuilder : public BaseOpBuilder {
+ public:
+  CastOpBuilder() : BaseOpBuilder("CastOpBuilder") {}
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(CastOpBuilder);
+
+ protected:
+  Status ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
+                       const NodeUnit& node_unit,
+                       const logging::Logger& logger,
+                       bool is_quantized_model,
+                       std::vector<std::string>& input_names,
+                       bool do_op_validation = false) const override ORT_MUST_USE_RESULT;
+
+  Status ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper,
+                                     const NodeUnit& node_unit,
+                                     std::vector<std::string>&& input_names,
+                                     const logging::Logger& logger,
+                                     bool is_quantized_model,
+                                     bool do_op_validation) const override ORT_MUST_USE_RESULT;
+
+};
+
+Status CastOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
+                                      const NodeUnit& node_unit,
+                                      const logging::Logger& logger,
+                                      bool is_quantized_model,
+                                      std::vector<std::string>& input_names,
+                                      bool do_op_validation) const {
+  ORT_UNUSED_PARAMETER(do_op_validation);
+  ORT_UNUSED_PARAMETER(is_quantized_model);  // Ignore in all backends. Cast should use same QNN types across backends.
+
+  const auto& inputs = node_unit.Inputs();
+  ORT_ENFORCE(inputs.size() == 1, "QNN Cast node must have a single input.");
+  const auto& input = inputs[0];
+
+  const auto& input_name = input.node_arg.Name();
+
+  if (qnn_model_wrapper.IsQnnTensorWrapperExist(input_name)) {
+    LOGS(logger, VERBOSE) << "Tensor already added, skip it: " << input_name;
+    input_names.push_back(input_name);
+    return Status::OK();
+  }
+
+  std::vector<uint8_t> unpacked_tensor;
+  bool is_initializer_input = qnn_model_wrapper.IsInitializerInput(input_name);
+  if (is_initializer_input) {
+    const auto& input_tensor = qnn_model_wrapper.GetInitializerTensors().at(input_name);
+    ORT_RETURN_IF_ERROR(onnxruntime::utils::UnpackInitializerData(*input_tensor, unpacked_tensor));
+  }
+
+  Qnn_TensorType_t tensor_type = GetInputTensorType(qnn_model_wrapper, input_name);
+  std::vector<uint32_t> input_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(input.node_arg, input_shape),
+                    "Cannot get shape for QNN Cast node's input.");
+
+  Qnn_DataType_t qnn_data_type = QNN_DATATYPE_UNDEFINED;
+  const auto* type_proto = input.node_arg.TypeAsProto();
+
+  ORT_RETURN_IF_ERROR(GetQnnDataType(false,  // Do not try to get the quantized type. HTP cast supports normal types.
```

Comment:
This helps and can make the model work. Wondering whether we can improve it further, for case like: uint8 input -> Cast (to float) -> Q -> DQ -> other nodes.
Is it possible to have node unit group to handle uin8 -> Cast -> Q all together? cast to quantized data directly instead of cast to float then Q node quantize it? vise versa.

---
In reply to: [1148918331](https://github.com/microsoft/onnxruntime/pull/15234#discussion_r1148918331) [](ancestors = 1148918331)

Reply:
This might be more work than may initially seem. Perhaps we can try to handle this as a follow-up PR?

Reply:
Sounds good.

---
In reply to: [1151188734](https://github.com/microsoft/onnxruntime/pull/15234#discussion_r1151188734) [](ancestors = 1151188734)